Update imghdr's what() to broadly accept readable binaries instead of BinaryIO

lawrence-law opened the pull request:
I'm proposing updating the type stubs for the stdlib `imghdr` so that it accepts `IO[bytes]` instead of `BinaryIO`. As per PEP 484, `BinaryIO` is a simple subtype of `IO[bytes]`. In `imghdr` we are merely reading a binary file to determine what image format it is and we don't strictly need `BinaryIO`.

Updating this will mean we can more flexibly use `imghdr.what()` with objects that are type annotated as `IO[bytes]`, for example in [`werkzeug.datastructures.FileStorage`](https://github.com/python/typeshed/blob/master/third_party/2and3/werkzeug/datastructures.pyi#L436). This change is also backwards compatible - i.e. anywhere we're already passing in `BinaryIO` to `imghdr.what()` will not be affected.

I have run `mypy_test.py`, `pytype_test.py`and `flake8` (on this file) without issue. I also ran `isort` to tidy up the imports.

srittau commented:
I don't think changing the type of `tests` this is right. Looking at [the implementation](/python/cpython/blob/8767ce92d24d3687405848442e6c67cf0af1c657/Lib/imghdr.py#L11) of `imghdr.what()`, it seems that the second argument to a test is either `None` (this case is actually missing from the annotation) or a file object as returned by `open()`. Therefore we should be as specific as possible, as tests can rely on that.

On the other hand, relaxing the `file` argument to `what()` makes sense, although using a custom protocol would be even better than using `IO[bytes]`.

lawrence-law commented:
> I don't think changing the type of `tests` this is right. Looking at [the implementation](/python/cpython/blob/8767ce92d24d3687405848442e6c67cf0af1c657/Lib/imghdr.py#L11) of `imghdr.what()`, it seems that the second argument to a test is either `None` (this case is actually missing from the annotation) or a file object as returned by `open()`. Therefore we should be as specific as possible, as tests can rely on that.
> 
> On the other hand, relaxing the `file` argument to `what()` makes sense, although using a custom protocol would be even better than using `IO[bytes]`.

Thanks for the feedback. I'll look into updating `tests` more faithfully.

Could you elaborate on using a custom protocol? Might just be terminology I'm not familiar with.

srittau commented:
We can use protocols to describe exactly what methods are used from an object. In our example:

```python
class _Readable(typing.Protocol):
    def tell(self) -> int: ...
    # also needs read() and seek()
```

This can then be used for `_File` instead of `BinaryIO` and allows to use any object that fulfills this protocol. If you are not comfortable trying this, I am also fine with using `IO[bytes]` for now.

lawrence-law commented:
Hi @srittau, just updated as per your feedback. I've added some notes against my own code if you could please provide some feedback. Thanks in advance!

lawrence-law commented on stdlib/2and3/imghdr.pyi:
```diff
@@ -1,16 +1,21 @@
+import sys
+from typing import Any, BinaryIO, Callable, List, Optional, Protocol, Text, Union, overload
+
+class _ReadableBinary(Protocol):
```
I thought `_ReadableBinary` was a better name given we're only interested in binary data here; after all it's image format checking.

lawrence-law commented on stdlib/2and3/imghdr.pyi:
```diff
@@ -1,16 +1,21 @@
+class _ReadableBinary(Protocol):
+    def tell(self) -> int: ...
+    def read(self, size: int = ...) -> bytes: ...
+    def seek(self, offset: int) -> Optional[int]: ...
```
We don't use the return of `seek` so I annotated this as `Optional[int]` - i.e. you don't have to return anything but if you do, an `int` is the right thing to return.

srittau replied:
I'd recommend to use `Any`, meaning "we don't care about the return code".

lawrence-law replied:
Will do.

lawrence-law commented on stdlib/2and3/imghdr.pyi:
```diff
@@ -1,16 +1,21 @@
+class _ReadableBinary(Protocol):
+    def tell(self) -> int: ...
+    def read(self, size: int = ...) -> bytes: ...
+    def seek(self, offset: int) -> Optional[int]: ...
```
I didn't add the `whence` parameter here as we see on `IO` since it's not used in `imghdr`. Hopefully this is the right way to implement a protocol - i.e. it doesn't seem to cause any issues when the class that we pass to `imghdr` does implement a `whence` parameter.

srittau replied:
Exactly.

srittau commented on stdlib/2and3/imghdr.pyi:
```diff
@@ -1,16 +1,21 @@
+
+class _ReadableBinary(Protocol):
+    def tell(self) -> int: ...
+    def read(self, size: int = ...) -> bytes: ...
```
Minor: You could leave out the requirement to have a default value, since it's not required by `what()`.

lawrence-law replied:
Done.

lawrence-law commented on stdlib/2and3/imghdr.pyi:
```diff
@@ -1,16 +1,21 @@
 @overload
 def what(file: Any, h: bytes) -> Optional[str]: ...
-tests: List[Callable[[bytes, BinaryIO], Optional[str]]]
+tests: List[Callable[[bytes, Optional[BinaryIO]], Optional[str]]]
```
So when you run `open(file, 'rb')` the type you get back is `BinaryIO` which is why I made it so here. I think it should be ok to keep it as such but keen to hear if there's something better to use here. FWIW, `f` isn't actually used in any of the test functions unless I'm mistaken.

lawrence-law commented:
> Thanks!

Thank you to you too! I learnt plenty :)
